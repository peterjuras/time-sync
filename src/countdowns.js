import { SECONDS, MINUTES, HOURS, DAYS, INTERVALS } from "./constants";
import { validateInterval } from "./lib/validation";
import { generateId } from "./lib/id";
import { getMs } from "./lib/get-ms";

const DEFAULT_COUNTDOWN_CONFIG = {
  interval: SECONDS
};

function validateCountdownConfig(countdownConfig) {
  if (!countdownConfig) {
    throw new Error("You must pass a valid countdown configuration object");
  }

  validateInterval(countdownConfig);

  if (!Number.isInteger(countdownConfig.until)) {
    throw new Error("until must be an integer");
  }
}

function convert(difference, interval) {
  const intervalValue = INTERVALS[interval];
  let result = difference / 1000;

  if (intervalValue >= INTERVALS[MINUTES]) {
    result /= 60;
  }

  if (intervalValue >= INTERVALS[HOURS]) {
    result /= 60;
  }

  if (intervalValue >= INTERVALS[DAYS]) {
    result /= 24;
  }

  return result;
}

function calculateTimeLeft(config) {
  // until is timestamp in milliseconds that should be reached
  const difference = config.until - Date.now();

  // Convert the difference to the specified interval
  const timeLeft = Math.ceil(convert(difference, config.interval));

  return timeLeft > 0 ? timeLeft : 0;
}

export function getTimeLeft(countdownConfig) {
  validateCountdownConfig(countdownConfig);

  const config = {
    ...DEFAULT_COUNTDOWN_CONFIG,
    ...countdownConfig
  };

  return calculateTimeLeft(config);
}

function getNextTickDelta({ ms, until }, time) {
  const timeMod = time % ms;
  const untilMod = until % ms;

  if (timeMod === untilMod) {
    return ms;
  }
  if (timeMod > untilMod) {
    return ms - timeMod;
  }
  return untilMod - timeMod;
}

function validateCreateCountdownArgs(callback, countdownConfig) {
  if (!callback) {
    throw new Error(
      "You need to provide a callback as the first argument to createCountdown"
    );
  }

  validateCountdownConfig(countdownConfig);
}

export class Countdowns {
  countdowns = {};

  createCountdown = (callback, countdownConfig) => {
    validateCreateCountdownArgs(callback, countdownConfig);

    const id = generateId();
    const newCountdown = {
      ...DEFAULT_COUNTDOWN_CONFIG,
      ...countdownConfig,
      callback,
      id
    };
    newCountdown.ms = getMs(newCountdown);

    const now = Date.now();
    const nextTickDelta = getNextTickDelta(newCountdown, now);
    if (newCountdown.until <= now) {
      throw new Error(
        "until property needs to be in the future. Please ensure you are using a millisecond timestamp!"
      );
    }
    // const nextTickDelta = nextTick - now;

    // eslint-disable-next-line no-param-reassign
    newCountdown.timeout = setTimeout(
      this.revalidate,
      nextTickDelta,
      // nextTick,
      newCountdown
    );

    this.countdowns[id] = newCountdown;

    return () => this.removeCountdown(id);
  };

  revalidate = countdown => {
    clearTimeout(countdown.timeout);

    const timeLeft = calculateTimeLeft(countdown);
    countdown.callback(timeLeft);

    if (timeLeft > 0) {
      // const nextTickDelta = getNextTick(countdown, Date.now()) - Date.now();
      const nextTickDelta = getNextTickDelta(countdown, Date.now());

      // eslint-disable-next-line no-param-reassign
      countdown.timeout = setTimeout(this.revalidate, nextTickDelta, countdown);
    } else {
      this.removeCountdown(countdown.id);
    }
  };

  removeCountdown = id => {
    const countdown = this.countdowns[id];
    if (!countdown) {
      return;
    }

    clearTimeout(countdown.timeout);
    delete this.countdowns[id];
  };

  stopAllCountdowns = () => {
    Object.keys(this.countdowns).forEach(id => {
      clearTimeout(this.countdowns[id].timeout);
    });
    this.countdowns = {};
  };
}
