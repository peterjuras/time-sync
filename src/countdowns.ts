import { Interval, IntervalMap } from "./constants";
import { validateInterval } from "./lib/validation";
import { generateId } from "./lib/id";
import { getMs } from "./lib/get-ms";

const DEFAULT_COUNTDOWN_CONFIG = {
  interval: Interval.SECONDS
};

interface CountdownConfig {
  until: number;
  interval?: Interval;
}

interface SafeCountdownConfig extends CountdownConfig {
  interval: Interval;
}

type CountdownCallback = (timeLeft: number) => void;

interface CountdownStoredConfig extends SafeCountdownConfig {
  id: number;
  callback: CountdownCallback;
  ms: number;
  timeout?: number;
}

function validateCountdownConfig(countdownConfig: CountdownConfig): void {
  if (!countdownConfig) {
    throw new Error("You must pass a valid countdown configuration object");
  }

  validateInterval(countdownConfig.interval);

  if (!Number.isInteger(countdownConfig.until)) {
    throw new Error("until must be an integer");
  }
}

function convert(difference: number, interval: Interval): number {
  const intervalValue = IntervalMap[interval];
  let result = difference / 1000;

  if (intervalValue >= IntervalMap[Interval.MINUTES]) {
    result /= 60;
  }

  if (intervalValue >= IntervalMap[Interval.HOURS]) {
    result /= 60;
  }

  if (intervalValue >= IntervalMap[Interval.DAYS]) {
    result /= 24;
  }

  return result;
}

function calculateTimeLeft(config: SafeCountdownConfig): number {
  // until is timestamp in milliseconds that should be reached
  const difference = config.until - Date.now();

  // Convert the difference to the specified interval
  const timeLeft = Math.ceil(convert(difference, config.interval));

  return timeLeft > 0 ? timeLeft : 0;
}

export function getTimeLeft(countdownConfig: CountdownConfig): number {
  validateCountdownConfig(countdownConfig);

  const config = {
    ...DEFAULT_COUNTDOWN_CONFIG,
    ...countdownConfig
  };

  return calculateTimeLeft(config);
}

function getNextTickDelta(
  { ms, until }: CountdownStoredConfig,
  time: number
): number {
  const timeMod = time % ms;
  const untilMod = until % ms;

  if (timeMod === untilMod) {
    return ms;
  }
  if (timeMod < untilMod) {
    return untilMod - timeMod;
  }
  return untilMod + ms - timeMod;
}

function validateCreateCountdownArgs(
  callback: CountdownCallback,
  countdownConfig: CountdownConfig
): void {
  if (!callback) {
    throw new Error(
      "You need to provide a callback as the first argument to createCountdown"
    );
  }

  validateCountdownConfig(countdownConfig);
}

export class Countdowns {
  private countdowns: {
    [key: string]: CountdownStoredConfig;
  } = {};

  public createCountdown = (
    callback: CountdownCallback,
    countdownConfig: CountdownConfig
  ) => {
    validateCreateCountdownArgs(callback, countdownConfig);

    const id = generateId();
    const countdownBase = {
      ...DEFAULT_COUNTDOWN_CONFIG,
      ...countdownConfig
    };
    const newCountdown: CountdownStoredConfig = {
      ...countdownBase,
      callback,
      id,
      ms: getMs(countdownBase.interval)
    };

    const now = Date.now();
    const nextTickDelta = getNextTickDelta(newCountdown, now);
    if (newCountdown.until <= now) {
      throw new Error(
        "until property needs to be in the future. Please ensure you are using a millisecond timestamp!"
      );
    }

    newCountdown.timeout = setTimeout(
      this.revalidate,
      nextTickDelta,
      newCountdown
    );

    this.countdowns[id] = newCountdown;

    return () => this.removeCountdown(id);
  };

  public stopAllCountdowns = () => {
    Object.keys(this.countdowns).forEach((id: string) => {
      clearTimeout(this.countdowns[id].timeout);
    });
    this.countdowns = {};
  };

  private revalidate = (countdown: CountdownStoredConfig) => {
    clearTimeout(countdown.timeout);

    const timeLeft = calculateTimeLeft(countdown);
    countdown.callback(timeLeft);

    if (timeLeft > 0) {
      const nextTickDelta = getNextTickDelta(countdown, Date.now());

      countdown.timeout = setTimeout(this.revalidate, nextTickDelta, countdown);
    } else {
      this.removeCountdown(countdown.id);
    }
  };

  private removeCountdown = (id: number) => {
    const countdown = this.countdowns[id];
    if (!countdown) {
      return;
    }

    clearTimeout(countdown.timeout);
    delete this.countdowns[id];
  };
}
