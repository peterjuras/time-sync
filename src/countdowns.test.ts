import TimeSync from "./index";
import { Countdowns } from "./countdowns";
import FakeTimers from "@sinonjs/fake-timers";

describe("#countdowns", () => {
  const instance = new Countdowns();
  let clock: FakeTimers.InstalledClock;

  beforeEach(() => {
    clock = FakeTimers.install({ now: 10001 });
  });

  afterEach(() => {
    clock.uninstall();
    instance.stopAllCountdowns();
  });

  describe("#createCountdown", () => {
    it("should be exported correctly", () => {
      expect(instance.createCountdown).toBeInstanceOf(Function);
    });

    it("should throw if no callback is provided", () => {
      expect(() =>
        instance.createCountdown(null as any, { until: 123 })
      ).toThrowErrorMatchingSnapshot();
    });

    it("should throw if for an unknown interval", () => {
      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: 123,
          interval: [] as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: 123,
          interval: {} as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: 123,
          interval: 123 as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: 123,
          interval: "" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: 123,
          interval: "a" as any,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it("should throw if for an invalid or missing until", () => {
      expect(() =>
        instance.createCountdown(jest.fn(), {} as any)
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: [] as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: {} as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: "" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: "a" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: "15" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.createCountdown(jest.fn(), {
          until: 250.291,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it("should add a second countdown by default", () => {
      const mock = jest.fn();
      instance.createCountdown(mock, { until: 13000 });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(998);
      expect(mock).toHaveBeenCalledTimes(0);

      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(1);

      clock.tick(999);
      expect(mock).toHaveBeenCalledTimes(1);

      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(2);

      clock.tick(10000);
      expect(mock).toHaveBeenCalledTimes(3);
    });

    it("should correctly callback for a second countdown", () => {
      const mock = jest.fn();
      instance.createCountdown(mock, {
        interval: TimeSync.SECONDS,
        until: 20 * 1000,
      });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(1000 * 3 - 2);

      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(8);
    });

    it("should correctly callback for a minute countdown", () => {
      const mock = jest.fn();
      instance.createCountdown(mock, {
        interval: TimeSync.MINUTES,
        until: 1000 * 3 * 60 + 10000,
      });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(60000 * 2 - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(2);
      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(1);
    });

    it("should correctly callback for an hour countdown", () => {
      const mock = jest.fn();
      instance.createCountdown(mock, {
        interval: TimeSync.HOURS,
        until: 1000 * 60 * 60 * 4 + 10000,
      });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(1000 * 60 * 60 * 2 - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(3);
      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(2);
    });

    it("should correctly callback for a day timer", () => {
      const mock = jest.fn();
      instance.createCountdown(mock, {
        interval: TimeSync.DAYS,
        until: 1000 * 60 * 60 * 4 * 24 + 10000,
      });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(1000 * 60 * 60 * 24 * 2 - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(3);
      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(2);
    });

    it("should work with a mix of countdowns", () => {
      const mockSecond = jest.fn();
      const mockSecondDelayed = jest.fn();
      const mockMinute = jest.fn();
      instance.createCountdown(mockSecond, { until: 2 * 1000 + 10000 });
      instance.createCountdown(mockSecondDelayed, { until: 2500 + 10000 });
      instance.createCountdown(mockMinute, {
        interval: TimeSync.MINUTES,
        until: 30 * 1000 + 10000,
      });

      clock.tick(1000 * 60);

      expect(mockSecond).toHaveBeenCalledTimes(2);
      expect(mockSecond.mock.calls).toMatchSnapshot();

      expect(mockSecondDelayed).toHaveBeenCalledTimes(3);
      expect(mockSecondDelayed.mock.calls).toMatchSnapshot();

      expect(mockMinute).toHaveBeenCalledTimes(1);
      expect(mockMinute.mock.calls).toMatchSnapshot();
    });

    it("adding a countdown should throw if the date is in the past", () => {
      expect(() =>
        instance.createCountdown(jest.fn(), { until: 0 })
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        instance.createCountdown(jest.fn(), { until: 1 })
      ).toThrowErrorMatchingSnapshot();
    });

    it("adding a countdown with an offset should be called back at the correct times", () => {
      const mock = jest.fn();
      instance.createCountdown(mock, { until: 2050 + 10000 });

      clock.tick(48);
      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(2);

      clock.tick(1000);
      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(1);

      clock.tick(999);
      expect(mock).toHaveBeenCalledTimes(2);

      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(3);
      expect(mock).toHaveBeenLastCalledWith(0);

      clock.tick(10000);
      expect(mock).toHaveBeenCalledTimes(3);
    });

    it("removing a countdown should not throw if it is called after all countdowns have been removed", () => {
      const mock = jest.fn();
      const ref = instance.createCountdown(mock, { until: 123 + 10000 });

      instance.stopAllCountdowns();
      ref();
    });

    it("removing a countdown should not throw if it is called multiple times", () => {
      const mock = jest.fn();
      const ref = instance.createCountdown(mock, { until: 123 + 10000 });

      ref();
      ref();
      ref();
    });
  });

  describe("#getTimeLeft", () => {
    it("should be exported correctly", () => {
      expect(TimeSync.getTimeLeft).toBeInstanceOf(Function);
    });

    it("should throw when no configuration object is passed", () => {
      expect(TimeSync.getTimeLeft).toThrowErrorMatchingSnapshot();
    });

    it("should give the correctly rounded value", () => {
      expect(TimeSync.getTimeLeft({ until: 11001 })).toBe(1);
      clock.tick(1250);
      expect(TimeSync.getTimeLeft({ until: 11001 })).toBe(0);
    });

    it("should work for defined countdown configurations", () => {
      const countdownConfig = {
        until: 1000 * 60 * 60 * 24 * 7,
        interval: TimeSync.DAYS,
      };
      expect(TimeSync.getTimeLeft(countdownConfig)).toBe(7);
      clock.tick(1000 * 60 * 60 * 24 * 5);
      expect(TimeSync.getTimeLeft(countdownConfig)).toBe(2);
    });

    it("should throw if there is no until property", () => {
      const countdownConfig: any = {
        interval: TimeSync.DAYS,
      };
      expect(() =>
        TimeSync.getTimeLeft(countdownConfig)
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe("#stopAllCountdowns", () => {
    it("should be exported correctly", () => {
      expect(instance.stopAllCountdowns).toBeInstanceOf(Function);
    });
  });
});
