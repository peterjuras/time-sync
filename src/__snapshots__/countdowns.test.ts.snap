// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#countdowns #createCountdown adding a countdown should throw if the date is in the past 1`] = `"until property needs to be in the future. Please ensure you are using a millisecond timestamp!"`;

exports[`#countdowns #createCountdown adding a countdown should throw if the date is in the past 2`] = `"until property needs to be in the future. Please ensure you are using a millisecond timestamp!"`;

exports[`#countdowns #createCountdown should throw if for an invalid or missing until 1`] = `"until must be an integer"`;

exports[`#countdowns #createCountdown should throw if for an invalid or missing until 2`] = `"until must be an integer"`;

exports[`#countdowns #createCountdown should throw if for an invalid or missing until 3`] = `"until must be an integer"`;

exports[`#countdowns #createCountdown should throw if for an invalid or missing until 4`] = `"until must be an integer"`;

exports[`#countdowns #createCountdown should throw if for an invalid or missing until 5`] = `"until must be an integer"`;

exports[`#countdowns #createCountdown should throw if for an invalid or missing until 6`] = `"until must be an integer"`;

exports[`#countdowns #createCountdown should throw if for an invalid or missing until 7`] = `"until must be an integer"`;

exports[`#countdowns #createCountdown should throw if for an unknown interval 1`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#countdowns #createCountdown should throw if for an unknown interval 2`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#countdowns #createCountdown should throw if for an unknown interval 3`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#countdowns #createCountdown should throw if for an unknown interval 4`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#countdowns #createCountdown should throw if for an unknown interval 5`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#countdowns #createCountdown should throw if no callback is provided 1`] = `"You need to provide a callback as the first argument to createCountdown"`;

exports[`#countdowns #createCountdown should work with a mix of countdowns 1`] = `
[
  [
    1,
  ],
  [
    0,
  ],
]
`;

exports[`#countdowns #createCountdown should work with a mix of countdowns 2`] = `
[
  [
    2,
  ],
  [
    1,
  ],
  [
    0,
  ],
]
`;

exports[`#countdowns #createCountdown should work with a mix of countdowns 3`] = `
[
  [
    0,
  ],
]
`;

exports[`#countdowns #getTimeLeft should throw if there is no until property 1`] = `"until must be an integer"`;

exports[`#countdowns #getTimeLeft should throw when no configuration object is passed 1`] = `"You must pass a valid countdown configuration object"`;
